pkg/codegen/dotnet/gen_program.go

```diff
@@ -0,0 +1,428 @@
+						}
+					}
+				})
+				g.Fprintf(w, "%s};\n", g.Indent)
+			})
+		}
+
+		// Close class constructor or Initialize
+		g.Fprintf(w, "%s}\n\n", g.Indent)
+
+		// Emit stack output properties
+		for _, n := range nodes {
+			switch n := n.(type) {
+			case *hcl2.OutputVariable:
+				g.genOutputProperty(w, n)
+			}
+		}
+	})
+	g.Fprint(w, "}\n")
+}
+
+func (g *generator) genNode(w io.Writer, n hcl2.Node) {
+	switch n := n.(type) {
+	case *hcl2.Resource:
+		g.genResource(w, n)
+	case *hcl2.ConfigVariable:
+		g.genConfigVariable(w, n)
+	case *hcl2.LocalVariable:
+		g.genLocalVariable(w, n)
+	case *hcl2.OutputVariable:
+		g.genOutputAssignment(w, n)
+	}
+}
+
+// requiresAsyncInit returns true if the program requires awaits in the code, and therefore an asynchronous
+// method must be declared.
+func requiresAsyncInit(r *hcl2.Resource) bool {
+	if r.Options == nil || r.Options.Range == nil {
+		return false
+	}
+
+	t := r.Options.Range.Type()
+	return model.ResolveOutputs(t) != t
+}
+
+// resourceTypeName computes the C# class name for the given resource.
+func (g *generator) resourceTypeName(r *hcl2.Resource) string {
+	// Compute the resource type from the Pulumi type token.
+	pkg, module, member, diags := r.DecomposeToken()
+	contract.Assert(len(diags) == 0)
+	if pkg == "pulumi" && module == "providers" {
+		pkg, module, member = member, "", "Provider"
+	}
+
+	namespaces := g.namespaces[pkg]
+	namespaceKey := strings.Replace(module, "/", ".", -1)
+	rootNamespace := namespaceName(namespaces, pkg)
+	namespace := namespaceName(namespaces, namespaceKey)
+
+	if namespace != "" {
+		namespace = "." + namespace
+	}
+
+	qualifiedMemberName := fmt.Sprintf("%s%s.%s", rootNamespace, namespace, Title(member))
+	return qualifiedMemberName
+}
+
+// functionName computes the C# namespace and class name for the given function token.
+func (g *generator) functionName(tokenArg model.Expression) (string, string) {
+	token := tokenArg.(*model.TemplateExpression).Parts[0].(*model.LiteralValueExpression).Value.AsString()
+	tokenRange := tokenArg.SyntaxNode().Range()
+
+	// Compute the resource type from the Pulumi type token.
+	pkg, module, member, diags := hcl2.DecomposeToken(token, tokenRange)
+	contract.Assert(len(diags) == 0)
+	namespaces := g.namespaces[pkg]
+	namespaceKey := strings.Replace(module, "/", ".", -1)
+	rootNamespace := namespaceName(namespaces, pkg)
+	namespace := namespaceName(namespaces, namespaceKey)
+
+	if namespace != "" {
+		namespace = "." + namespace
+	}
+
+	return rootNamespace, fmt.Sprintf("%s%s.%s", rootNamespace, namespace, Title(member))
+}
+
+// argumentTypeName computes the C# argument class name for the given expression and model type.
+func (g *generator) argumentTypeName(expr model.Expression, destType model.Type) string {
+	if schemaType, ok := hcl2.GetSchemaForType(destType.(model.Type)); ok {
+		if objType, ok := schemaType.(*schema.ObjectType); ok {
+			token := objType.Token
```

Comment:
Using early returns here is a bit more idiomatic:

```suggestion
	schemaType, ok := hcl2.GetSchemaForType(destType.(model.Type))
	if !ok {
		return ""
	}
	objType, ok := schemaType.(*schema.ObjectType)
	if !ok {
		return ""
	}
	
	token := objType.Token
```